clover.rb

```diff
@@ -3,38 +3,688 @@
+      headers.delete("content-security-policy") if json
+    end
+  end
+
+  plugin :all_verbs
+  plugin :assets, js: "app.js", css: "app.css", css_opts: {style: :compressed, cache: false}, timestamp_paths: true
+  plugin :disallow_file_uploads
+  plugin :flash
+  plugin :h
+  plugin :hash_branches
+  plugin :hash_branch_view_subdir
+  plugin :Integer_matcher_max
+  plugin :json
+  plugin :json_parser
+  plugin :public
+  plugin :render, escape: true, layout: "./layouts/app", template_opts: {chain_appends: true, freeze: true, skip_compiled_encoding_detection: true}
+  plugin :request_headers
+  plugin :typecast_params_sized_integers, sizes: [64], default_size: 64
+
+  plugin :sessions,
+    key: "_Clover.session",
+    cookie_options: {secure: !(Config.development? || Config.test?)},
+    secret: Config.clover_session_secret
+
+  plugin :route_csrf do |token|
+    flash["error"] = "An invalid security token submitted with this request, please try again"
+    return redirect_back_with_inputs
+  end
+
+  plugin :content_security_policy do |csp|
+    csp.default_src :none
+    csp.style_src :self, "https://cdn.jsdelivr.net/npm/flatpickr@4.6.13/dist/flatpickr.min.css"
+    csp.img_src :self, "data: image/svg+xml"
+    csp.form_action :self, "https://checkout.stripe.com"
+    csp.script_src :self, "https://cdn.jsdelivr.net/npm/jquery@3.7.0/dist/jquery.min.js", "https://cdn.jsdelivr.net/npm/dompurify@3.0.5/dist/purify.min.js", "https://cdn.jsdelivr.net/npm/flatpickr@4.6.13/dist/flatpickr.min.js"
+    csp.connect_src :self
+    csp.base_uri :none
+    csp.frame_ancestors :none
+  end
+
+  logger = if ENV["RACK_ENV"] == "test"
+    Class.new {
+      def write(_)
+      end
+    }.new
+  else
+    # :nocov:
+    $stderr
+    # :nocov:
+  end
+  plugin :common_logger, logger
+
+  # XXX: Temporary unless PR 2084 is merged
+  # :nocov:
+  if Config.production?
+    def api?
+      return @is_api if defined?(@is_api)
+      @is_api = env["HTTP_HOST"]&.start_with?("api.")
+    end
+  else
+    def api?
+      return @is_api if defined?(@is_api)
+      @is_api = env["HTTP_HOST"]&.start_with?("api.") || env["PATH_INFO"].start_with?("/api")
+    end
+  end
+  # :nocov:
+
+  def runtime?
+    return @is_runtime if defined?(@is_runtime)
+    @is_runtime = env["PATH_INFO"].start_with?("/runtime")
+  end
+
+  def web?
+    return @is_web if defined?(@is_web)
+    @is_web = !api? && !runtime?
+  end
+
+  def get_jwt_payload(request)
+    return unless (v = request.env["HTTP_AUTHORIZATION"])
+    jwt_token = v.sub(%r{\ABearer:?\s+}, "")
+    begin
+      JWT.decode(jwt_token, Config.clover_runtime_token_secret, true, {algorithm: "HS256"})[0]
+    rescue JWT::DecodeError
+    end
+  end
+
+  def csrf_tag(*)
+    render("components/form/hidden", locals: {name: csrf_field, value: csrf_token(*)})
+  end
+
+  def redirect_back_with_inputs
+    flash["old"] = request.params
+    request.redirect env["HTTP_REFERER"] || "/"
+  end
```

Comment:
Helpers like `csrf_tag`, `redirect_back_with_inputs` specifically for web, and `get_jwt_payload` specifically for runtime. We could consider moving these helpers to the `routes/helpers` directory for better isolation.

Reply:
OK, I'll push a commit to do that.  I'll move the custom methods to helpers, so that clover.rb just has the plugin configurations and the routes.